jlm/llvm/opt/reduction.cpp

```diff
@@ -4,132 +4,233 @@
 {
-  auto nf = LoadNonVolatileOperation::GetNormalForm(&graph);
-  nf->set_mutable(true);
-  nf->set_load_mux_reducible(true);
-  nf->set_load_store_reducible(true);
-  nf->set_load_alloca_reducible(true);
-  nf->set_multiple_origin_reducible(true);
-  nf->set_load_store_state_reducible(true);
-  nf->set_load_load_state_reducible(true);
+  const auto & graph = rvsdgModule.Rvsdg();
+
+  Statistics_ = Statistics::Create(rvsdgModule.SourceFileName());
+  Statistics_->Start(graph);
+
+  ReduceNodesInRegion(graph.GetRootRegion());
+
+  Statistics_->End(graph);
+  statisticsCollector.CollectDemandedStatistics(std::move(Statistics_));
 }
 
-static void
-enable_gamma_reductions(rvsdg::Graph & graph)
+void
+NodeReduction::ReduceNodesInRegion(rvsdg::Region & region)
 {
-  auto nf = rvsdg::GammaOperation::normal_form(&graph);
-  nf->set_mutable(true);
-  nf->set_predicate_reduction(true);
-  // set_control_constante_reduction cause a PHI node input type error
-  // github issue #303
-  nf->set_control_constant_reduction(false);
+  bool reductionPerformed;
+  size_t numIterations = 0;
+  do
+  {
+    numIterations++;
+    reductionPerformed = false;
+
+    for (const auto node : rvsdg::topdown_traverser(&region))
+    {
+      if (const auto structuralNode = dynamic_cast<rvsdg::StructuralNode *>(node))
+      {
+        reductionPerformed |= ReduceStructuralNode(*structuralNode);
+      }
+      else if (rvsdg::is<rvsdg::SimpleOperation>(node))
+      {
+        reductionPerformed |= ReduceSimpleNode(*node);
+      }
+      else
+      {
+        JLM_UNREACHABLE("Unhandled node type.");
+      }
+    }
+
+    if (reductionPerformed)
+    {
+      // Let's remove all dead nodes in this region to avoid reductions on
+      // dead nodes in the next iteration.
+      region.prune(false);
+    }
+  } while (reductionPerformed);
+
+  Statistics_->AddIteration(region, numIterations);
 }
 
-static void
-enable_unary_reductions(rvsdg::Graph & graph)
+bool
+NodeReduction::ReduceStructuralNode(rvsdg::StructuralNode & structuralNode)
 {
-  auto nf = jlm::rvsdg::unary_op::normal_form(&graph);
-  // set_mutable generates incorrect output for a number of
-  // llvm suite tests when used in combination with other
-  // optimizations than the set_reducible
-  nf->set_mutable(false);
-  // set_reducible generates incorrect output for 18 llvm suite tests
-  // github issue #304
-  nf->set_reducible(false);
+  bool reductionPerformed = false;
+
+  // Reduce structural nodes
+  if (is<rvsdg::GammaOperation>(&structuralNode))
+  {
+    reductionPerformed |= ReduceGammaNode(structuralNode);
+  }
+
+  if (reductionPerformed)
+  {
+    // We can not go through the subregions as the structural node might already have been removed.
+    return true;
+  }
+
+  // Reduce all nodes in the subregions
+  for (size_t n = 0; n < structuralNode.nsubregions(); n++)
+  {
+    const auto subregion = structuralNode.subregion(n);
+    ReduceNodesInRegion(*subregion);
```

Comment:
Should the result of this function call be logical OR-ed into `reductionPerformed`?